packages/stack/contracts-manager/test/contracts-manager.spec.js

```diff
@@ -0,0 +1,199 @@
+import ContractsManager from '../src';
+import Contract from '../src/contract';
+
+import sinon from 'sinon';
+import assert from 'assert';
+import { fakeEmbark } from 'embark-testing';
+
+const { embark, plugins } = fakeEmbark();
+
+// Due to our `DAPP_PATH` dependency in `embark-utils` `dappPath()`, we need to
+// ensure that this environment variable is defined.
+process.env.DAPP_PATH = 'something';
+
+describe('stack/contracts-manager', () => {
+
+  let contractsManager;
+
+  beforeEach(() => {
+    contractsManager = new ContractsManager(embark, { plugins });
+  });
+
+  afterEach(() => {
+    embark.teardown();
+    sinon.restore();
+  });
+
+  it('should add contract objects', async () => {
+    const testContract = {
+      className: 'TestContract'
+    };
+
+    await embark.events.request2('contracts:add', testContract);
+    const contract = await embark.events.request2('contracts:contract', testContract.className);
+    assert.equal(contract.className, testContract.className);
+  });
+
+  it('should return list of added contract objects', async () => {
+    const testContract = {
+      className: 'TestContract'
+    };
+
+    await embark.events.request2('contracts:add', testContract);
+    const list = await embark.events.request2('contracts:list');
+    assert.equal(list.length, 1);
+    assert.equal(list[0].className, testContract.className);
+  });
+
+  it('should build contracts', async (done) => {
+
+    const contractsConfig = {
+      contracts: {
+        deploy: {
+          TestContract: {}
+        }
+      }
+    };
+
+    const compiledContracts = {
+      TestContract: {
+        // `compiledContract` can't be an empty object otherwise
+        // contracts-manager will exclude it from processing, resulting
+        // in `TestContract` not being managed.
+        code: 'some code',
+      }
+    };
+
+    const beforeBuildAction = sinon.spy((params, cb) => { cb(null, params); });
+    embark.registerActionForEvent('contracts:build:before', beforeBuildAction);
+
+    embark.events.request('contracts:build', contractsConfig, compiledContracts, (err, result) => {
```

Comment:
It would be nice to split this in to two tests:
1. Test that `contracts:build` command handler was registered (in the constructor section, see comment above) using `contractsManager.events.assert.commandHandlerRegistered('contracts:add');`
2. Test the `buildContracts` method separately and call it directly in the unit test, without going through the event bus, ie:
```
// setup
const beforeBuildAction = sinon.fake.yields(null, params);
contractsManager.plugins.registerActionForEvent('contracts:build:before', beforeBuildAction);
contractsManager.buildContracts(contractsConfig, compiledContracts, (err, result) => {
  // assertions
  contractsManager.plugins.assert.actionForEventCalled('contracts:build:before', action);
  assert.ok(result.TestContract instanceof Contract);
  done();
});
```

Reply:
Okay, will update the tests. So far I have been testing through the event bus APIs as those are the APIs that are considered "public API". In fact, I would even argue that, due to that system, all methods could be made private on modules like this one.

Anyways, just thinking out loud here!